src/context.ts

```diff
@@ -14,27 +14,24 @@ import {
   useStore,
 } from 'zustand'
 
-export type UseContextStore<T extends State> = {
-  (): T
-  <U>(selector: StateSelector<T, U>, equalityFn?: EqualityChecker<U>): U
-}
+export type UseContextStore<S> = <U = ExtractState<S>>(
+  selector?: (state: ExtractState<S>) => U,
+  equals?: (a: U, b: U) => boolean
+) => U
+
+type ExtractState<S> = S extends { getState: () => infer T } ? T : never
 
-function createContext<
-  TState extends State,
-  CustomStoreApi extends StoreApi<TState> = StoreApi<TState>
->() {
-  const ZustandContext = reactCreateContext<CustomStoreApi | undefined>(
-    undefined
-  )
+function createContext<S extends StoreApi<State>>() {
```

Comment:
This type change may break existing app code. Hope we can give migration path in v3.7.0. Can we do that in #715?

Reply:
We can probably add an overload of the older type, so yes we can have a migration path in v3.7.0

Reply:
We can keep it but I think the newer version is better because it encourages users to pass the store type instead of the state type. If they pass the state type, then when they use `useStoreApi()` they'd get the default store instead of the one they would have passed to the provider with middlewares. I hope I've made myself clear. Let me know if you want me to revert it.

Reply:
Yes, i agree the new one is better.